pkg/streamingpromql/types/stats.go

```diff
@@ -21,6 +24,75 @@ type QueryStats struct {
 	// For example, if a query is running with a step of 30s with a range vector selector with range 45s,
 	// then samples in the overlapping 15s are counted twice.
 	TotalSamples int64
+
+	// TotalSamplesPerStep represents the total number of samples scanned
+	// per step while evaluating a query. Each step should be identical to the
+	// TotalSamples when a step is run as an instant query, which means
+	// we intentionally do not account for optimizations that happen inside the
+	// range query engine that reduce the actual work that happens.
+	TotalSamplesPerStep []int64
+
+	EnablePerStepStats bool
+	timeRange          QueryTimeRange
+
+	memoryConsumptionTracker *limiter.MemoryConsumptionTracker
+}
+
+func NewQueryStats(timeRange QueryTimeRange, enablePerStepStats bool, memoryConsumptionTracker *limiter.MemoryConsumptionTracker) (*QueryStats, error) {
+	qs := &QueryStats{
+		EnablePerStepStats:       enablePerStepStats,
+		timeRange:                timeRange,
+		memoryConsumptionTracker: memoryConsumptionTracker,
+	}
+	if enablePerStepStats {
+		s, err := Int64SlicePool.Get(qs.timeRange.StepCount, qs.memoryConsumptionTracker)
+		if err != nil {
+			return nil, fmt.Errorf("failed to allocate TotalSamplesPerStep: %w", err)
+		}
+		qs.TotalSamplesPerStep = s[:qs.timeRange.StepCount]
+	}
+	return qs, nil
+}
+
+// IncrementSamplesAtStep increments the total samples count. Use this if you know the step index.
+func (qs *QueryStats) IncrementSamplesAtStep(i int, samples int64) {
+	if qs == nil {
+		return
+	}
+	qs.TotalSamples += samples
+
+	if qs.TotalSamplesPerStep != nil {
+		qs.TotalSamplesPerStep[i] += samples
+	}
+}
+
+// IncrementSamplesAtTimestamp increments the total samples count. Use this if you only have the corresponding step
+// timestamp.
+func (qs *QueryStats) IncrementSamplesAtTimestamp(t, samples int64) {
+	if qs == nil {
+		return
+	}
+	qs.TotalSamples += samples
+
+	if qs.TotalSamplesPerStep != nil {
+		qs.TotalSamplesPerStep[qs.timeRange.PointIndex(t)] += samples
+	}
+}
+
+// Clear resets the TotalSamples counter to 0 and zeroes out all entries in the
+// TotalSamplesPerStep slice, preserving its length and capacity for reuse.
+func (qs *QueryStats) Clear() {
+	qs.TotalSamples = 0
+	if qs.TotalSamplesPerStep != nil {
+		clear(qs.TotalSamplesPerStep)
+	}
```

Comment:
I don't think we need the `if` here, [the spec for `clear`](https://go.dev/ref/spec#Clear) says it'll work fine with `nil` slices:

>If the map or slice is `nil`, `clear` is a no-op.